### PR TITLE
FaissIndexer example usage

### PR DESCRIPTION
**Pull request by JoanFM**
This PR adds an example that uses FaissIndexer as a vector database to build and query vectors.

It requires to push a new version of  **jinaai/hub.executors.indexers.vector.faiss:latest**

It requires the changes in https://github.com/jina-ai/jina/pull/621 to work.

Since new comments in the PR, https://github.com/jina-ai/jina/pull/635 is necessary to work

### CODE REVIEW

File: faiss-search/app.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+@click.option('--task', '-t')
+@click.option('--batch_size', '-n', default=50)
+@click.option('--top_k', '-k', default=5)
+def main(task, batch_size, top_k):
+    os.environ['TMP_WORKSPACE'] = get_random_ws(os.environ['TMP_DATA_DIR'])
+    if task == 'index':
+        data_path = os.path.join(os.environ['TMP_DATA_DIR'], 'siftsmall_base.fvecs')
+        flow = Flow().load_config('flow-index.yml')
+        with flow.build() as fl:
+            fl.index(read_data(data_path), batch_size=batch_size)
```
Comment (nan-wang) [inline faiss-search/app.py:89]: `index_ndarray` is dedicated to indexing arrays.
```suggestion
            fl.index_ndarray(read_data(data_path), batch_size=batch_size)
```

File: faiss-search/app.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    elif task == 'query':
+        data_path = os.path.join(os.environ['TMP_DATA_DIR'], 'siftsmall_query.fvecs')
+        flow = Flow().load_config('flow-query.yml')
+        with flow.build() as fl:
+            ppr = lambda x: save_topk(x, os.path.join(os.environ['TMP_DATA_DIR'], 'query_results.txt'), top_k)
+            fl.search(buffer=read_data(data_path), callback=ppr, top_k=top_k)
```
Comment (nan-wang) [inline faiss-search/app.py:95]: `callback` will be deprecated
```suggestion
            fl.search_ndarray(read_data(data_path), output_fn=ppr, top_k=top_k)
```

**Comment by nan-wang** [comment]
@JoanFM BTW, it would be great if you can compare the query performances of FAISS with that of Numpy. Han has given a trial with Helloworld example. We could also test Helloworld with faiss.

Reference:
[https://github.com/jina-ai/jina/issues/619](https://github.com/jina-ai/jina/issues/619)

**Comment by JoanFM** [comment]
Hey @nan-wang, thank you very much for the comments.

With respect to 


> @JoanFM BTW, it would be great if you can compare the query performances of FAISS with that of Numpy. Han has given a trial with Helloworld example. We could also test Helloworld with faiss.
> 
> Reference:
> [jina-ai/jina#619](https://github.com/jina-ai/jina/issues/619)

I think it is a good idea. Now I am planning to write some stress tests to analyze better performance and quality tradeoffs of FAISS, I plan it to be also usable for other indexers so that we can compare.

I also think adding FaissIndexer to Hello-World makes a lot of sense under the scope of https://github.com/jina-ai/jina/issues/619. The only issue is what images we use for training the index? would it make sense to add an option instead of asking for a train_filepath, have the possibility to use the same numpy index and take some percentage of it for training? I think it may make early adoption of FAISS as vector database easier.

File: faiss-search/yaml/customized_executors.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    def craft(self, blob, *args, **kwargs):
+        return [dict(blob=blob)]
+
+
+class MyEncoder(BaseNumericEncoder):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def encode(self, data: 'np.ndarray', *args, **kwargs):
+        return data
```
[the remark this answers is not in the document]
Comment (JoanFM) [reply faiss-search/yaml/customized_executors.py:21]: Should I better add a UnaryEncoderDriver? it would be coherent with having a UnarySegmenterDriver?

**Comment by JoanFM** [comment]
@nan-wang changes implementing the comments in core requests are in https://github.com/jina-ai/jina/pull/635

**Comment by JoanFM** [comment]
@nan-wang requested changes have been applied, could we publish a new image of the faiss indexer before merging this example?

**Pull request merged**